src/tactic/positivity.lean

```diff
@@ -363,4 +361,45 @@ meta def positivity_abs : expr → tactic strictness
   nonnegative <$> mk_app ``abs_nonneg [a] -- else report nonnegativity
 | _ := failed
 
+/-- Extension for the `positivity` tactic: Any element of a canonically ordered additive monoid is
+nonnegative. -/
+@[positivity]
+meta def positivity_canon : expr → tactic strictness
+| `(%%a) := nonnegative <$> mk_app ``zero_le [a]
```

Comment:
This seems a bit suspicious, especially with the failing test cases. On the other hand, from the code in the `positivity` attribute it looks like failing tactics are skipped. And I don't think `mk_app` would succeed with a missing instance, so probably the error is somewhere else.

I'm still not convinced it's a good idea having a tactic that is expected to try and fail to synthesize a `canonically_ordered_add_monoid` instance on all expressions (especially considering the `simp` `subsingleton` issue). Or is the search space small enough that this does not slow down the tactic appreciably?

Reply:
I agree; positivity extenstions should always start with a head check of some kind so that it can be skipped on most expressions. If it's universally applicable it should go in the core tactic.

As for whether this is too many typeclass searches, I'm not sure; it's one per "positivity locus" (i.e. all the subexpressions of the goal but not things like `(+)` in `a + b + c`, only `a` and `b` and `c` and `a + b`). `simp` has things quite a bit worse because it also has to consider subexpressions after replacement, so it's not bounded by the size of the initial goal.

Reply:
How does one add this to the core tactic?

Reply:
Okay, I tried something and it seems to work. I don't know whether I placed it sensibly, though.